File: go/parsers/publish.go
package parsers

import (
	"google.golang.org/protobuf/proto"

	"github.com/xconnio/wampproto-go/messages"
	"github.com/xconnio/wampproto-go/serializers"
	"github.com/xconnio/wampproto-protobuf/go/gen"
)

type Publish struct {
	gen *gen.Publish
	ex  *PayloadExpander
}

func NewPublishFields(gen *gen.Publish, payload []byte) messages.PublishFields {
	return &Publish{
		gen: gen,
		ex:  &PayloadExpander{payload: payload, serializer: gen.GetPayloadSerializerId()},
	}
}

func (p *Publish) RequestID() uint64 {
	return p.gen.GetRequestId()
}

func (p *Publish) Options() map[string]any {
	var details map[string]any

	if !p.gen.ExcludeMe {
		setDetail(&details, "exclude_me", false)
	}

	if p.gen.Acknowledge {
		setDetail(&details, "acknowledge", true)
	}

	return details
}

func (p *Publish) Topic() string {
	return p.gen.GetTopic()
}

func (p *Publish) Args() []any {
	return p.ex.Args()
}

func (p *Publish) KwArgs() map[string]any {
	return p.ex.Kwargs()
}

func (p *Publish) PayloadIsBinary() bool {
	return true
}

func (p *Publish) Payload() []byte {
	return p.ex.Payload()
}

func (p *Publish) PayloadSerializer() uint64 {
	return p.gen.GetPayloadSerializerId()
}

func PublishToProtobuf(publish *messages.Publish) ([]byte, error) {
	msg := &gen.Publish{
		RequestId: publish.RequestID(),
		Topic:     publish.Topic(),
	}
	acknowledge, ok := publish.Options()["acknowledge"].(bool)
	if ok {
		msg.Acknowledge = acknowledge
	}

	excludeMe, ok := publish.Options()["exclude_me"].(bool)
	if ok {
		msg.ExcludeMe = excludeMe
	}

	payloadSerializer := selectPayloadSerializer(publish.Options())
	msg.PayloadSerializerId = payloadSerializer

	var payload []byte
	if publish.PayloadIsBinary() {
		payload = publish.Payload()
	} else {
		var err error
		payload, err = serializers.SerializePayload(payloadSerializer, publish.Args(), publish.KwArgs())
		if err != nil {
			return nil, err
		}
	}

	data, err := proto.Marshal(msg)
	if err != nil {
		return nil, err
	}

	return PrependHeader(messages.MessageTypePublish, data, payload), nil
}

func ProtobufToPublish(data, payload []byte) (*messages.Publish, error) {
	msg := &gen.Publish{}
	err := proto.Unmarshal(data, msg)
	if err != nil {
		return nil, err
	}

	return messages.NewPublishWithFields(NewPublishFields(msg, payload)), nil
}

File: python/wampprotobuf/parsers/publish.py
from typing import Any

from wampproto.messages import Message
from wampproto.messages.publish import IPublishFields, Publish

from wampprotobuf.gen import publish_pb2
from wampprotobuf.parsers import helpers


class PublishFields(IPublishFields):
    def __init__(self, msg: publish_pb2.Publish):
        super().__init__()
        self._msg = msg
        self._args = None
        self._kwargs = None
        self._unpacked = False

    @property
    def request_id(self) -> int:
        return self._msg.request_id

    @property
    def uri(self) -> str:
        return self._msg.topic

    def unpack(self):
        try:
            self._unpacked = True
            args, kwargs = helpers.from_cbor_payload(self._msg.payload)
            self._args = args
            self._kwargs = kwargs
        except Exception as e:
            print(f"error parsing CBOR payload: {e}")

    @property
    def args(self) -> list[Any] | None:
        if not self._unpacked:
            self.unpack()

        return self._args

    @property
    def kwargs(self) -> dict[str, Any] | None:
        if not self._unpacked:
            self.unpack()

        return self._kwargs

    @property
    def options(self) -> dict[str, Any]:
        return {}

    def payload_is_binary(self) -> bool:
        return True

    @property
    def payload(self) -> bytes | None:
        return self._msg.payload

    @property
    def payload_serializer(self) -> int:
        return self._msg.payload_serializer


def from_protobuf(payload: bytes) -> Message:
    result = publish_pb2.Publish()
    result.ParseFromString(payload)

    return Publish(PublishFields(result))


def to_protobuf(publish: Publish) -> bytes:
    result = publish_pb2.Publish()
    result.request_id = publish.request_id
    result.topic = publish.uri

    if publish.payload_is_binary():
        result.payload = publish.payload
        result.payload_serializer = publish.payload_serializer
    else:
        try:
            payload = helpers.to_cbor_payload(publish.args, publish.kwargs)
        except Exception as e:
            raise Exception(f"error in serialization to cbor: {e}")

        serializer = 1

        result.payload = payload
        result.payload_serializer = serializer

    return result.SerializeToString()
